### ISSUE TITLE
RADOS lots of small files(4-10KB),but waste much more space 

### ISSUE THREAD

**Issue by ijackcy** [rook/rook#6032]
rook:1.3
k8s:1.18.6
one master and three slaver
osd:5

[root@rook-ceph-tools-75f575489-m2t56 /]# rados df
POOL_NAME                      USED OBJECTS CLONES COPIES MISSING_ON_PRIMARY UNFOUND DEGRADED RD_OPS      RD WR_OPS     WR USED COMPR UNDER COMPR 
.rgw.root                   1.9 MiB      16      0     32                  0       0        0     76  76 KiB     32 23 KiB        0 B         0 B 
my-store.rgw.buckets.data       0 B       0      0      0                  0       0        0      0     0 B      0    0 B        0 B         0 B 
my-store.rgw.buckets.index      0 B       1      0      2                  0       0        0      7   7 KiB      2  1 KiB        0 B         0 B 
my-store.rgw.buckets.non-ec     0 B       0      0      0                  0       0        0      0     0 B      0    0 B        0 B         0 B 
my-store.rgw.control            0 B       8      0     16                  0       0        0      0     0 B      0    0 B        0 B         0 B 
my-store.rgw.log            128 KiB     177      0    354                  0       0        0    526 350 KiB    353  1 KiB        0 B         0 B 
my-store.rgw.meta           512 KiB       5      0     10                  0       0        0      5   5 KiB     10  5 KiB        0 B         0 B 

then i create bucket,  replicated size:2 , and upload  lots of small files, the size of the files many be 4KB-16KB, the total size is  61MB. but i found the actual size is 607MB, and many space was wasted.

4868*2*64KB=608M,nearly the actual size 607.

[root@rook-ceph-tools-75f575489-m2t56 /]# rados df
POOL_NAME                      USED OBJECTS CLONES COPIES MISSING_ON_PRIMARY UNFOUND DEGRADED RD_OPS      RD WR_OPS      WR USED COMPR UNDER COMPR 
.rgw.root                   1.9 MiB      16      0     32                  0       0        0     76  76 KiB     32  23 KiB        0 B         0 B 
my-store.rgw.buckets.data   607 MiB    4868      0   9736                  0       0        0      0     0 B  58386  61 MiB        0 B         0 B 
my-store.rgw.buckets.index      0 B       1      0      2                  0       0        0  29224  29 MiB  14605 9.5 MiB        0 B         0 B 
my-store.rgw.buckets.non-ec     0 B       0      0      0                  0       0        0      0     0 B      0     0 B        0 B         0 B 
my-store.rgw.control            0 B       8      0     16                  0       0        0      0     0 B      0     0 B        0 B         0 B 
my-store.rgw.log            128 KiB     177      0    354                  0       0        0    955 779 KiB    639   1 KiB        0 B         0 B 
my-store.rgw.meta           512 KiB       5      0     10                  0       0        0     11  11 KiB     16   8 KiB        0 B         0 B 

i found the default bluestore min alloc size is 64KB, but the size of small files are 4KB-16KB, and i  tried to modify the " bluestore min alloc size " to 16KB, also i delete the rook-ceph, and install again.

apiVersion: v1
kind: ConfigMap
metadata:
  name: rook-config-override
  namespace: rook-ceph
data:
  config: |
    [global]
    bluestore_min_alloc_size = 16384

and i exec the pod of osd1, 
[root@rook-ceph-osd-0-659bd5d7f4-pn6wc /]# cat /etc/ceph/..2020_08_11_05_51_57.356566463/ceph.conf 
[global]
bluestore_min_alloc_size = 16384

but it didn't work, and still waste much space.  how can i do ?

**Comment by satoru-takeuchi** [comment]
It looks not a Rook's problem, but a Ceph's one since Rook don't deal with how RADOS works. It's better to ask in ceph-users ML.

https://ceph.io/resources/

**Comment by ijackcy** [comment]
i setup the ceph by ceph-deploy.  and the ceph version is v14.2.10,i modify the /etc/ceph/ceph.conf .and it works well.
i can modify bluestore_min_alloc_size

**Comment by satoru-takeuchi** [comment]
I'll try to reproduce your problem next week.

**Comment by satoru-takeuchi** [comment]
@ijackcy I encountered the same problem. Although /etc/ceph/ceph.conf looks fine, this configuration is not reflected to Ceph cluster.

I'll research why it happens.

**Comment by satoru-takeuchi** [comment]
@travisn 

https://github.com/rook/rook/issues/6140
> For the osd memory target, it's expected to be set internally by rook and ceph with the resource limits in the cluster CR.

Is this description also applied to bluestore_min_alloc_size? I couldn't find this symbol in Rook's source.

I confirmed the following things in master, v1.3.0, and v1.2.7:
- "bluestore_min_alloc_size = 16384" is set in /etc/ceph/ceph.conf in the rook-ceph pods
- `ceph config` shows the bluestore_min_alloc_size = 0.

```console
$ kubectl -n rook-ceph exec rook-ceph-tools-55d5c49f79-58fsj -- ceph config dump
WHO    MASK LEVEL    OPTION                            VALUE RO
global      advanced mon_allow_pool_delete             true
global      advanced rbd_default_features              3
  mgr       advanced mgr/orchestrator_cli/orchestrator rook  *
$ kubectl -n rook-ceph exec rook-ceph-tools-55d5c49f79-58fsj -- ceph config get osd.0 bluestore_min_alloc_size
0
```

- This value can be changed vio `ceph config set`

```console
$ kubectl -n rook-ceph exec rook-ceph-tools-55d5c49f79-58fsj -- ceph config set osd bluestore_min_alloc_size 16384
$ kubectl -n rook-ceph exec rook-ceph-tools-55d5c49f79-58fsj -- ceph config dump
WHO    MASK LEVEL    OPTION                            VALUE RO
global      advanced mon_allow_pool_delete             true
global      advanced rbd_default_features              3
  mgr       advanced mgr/orchestrator_cli/orchestrator rook  *
  osd       advanced bluestore_min_alloc_size          16384 *
```

**Comment by travisn** [comment]
> @travisn
> 
> #6140
> 
> > For the osd memory target, it's expected to be set internally by rook and ceph with the resource limits in the cluster CR.
> 
> Is this description also applied to bluestore_min_alloc_size? I couldn't find this symbol in Rook's source.

No, the resource limits only apply to the memory limits. The `bluestore_min_alloc_size` is not changed by rook unless you specify it in the ceph.conf overrides.

**Comment by ijackcy** [comment]
@satoru-takeuchi  thank you very much!
 i tried many times , but it still didn't work well.
then i tried to install ceph by ceph-deploy,not rook. and modify /etc/ceph/ceph.conf, it work very well.
but in the k8s environment, rook is the best choice. and i still wish rook can solve the problem.

**Comment by ijackcy** [comment]
@travisn 
thank you!

1、I created the file  ceph.yaml  and add bluestore_min_alloc_size = 16384
apiVersion: v1
kind: ConfigMap
metadata:
name: rook-config-override
namespace: rook-ceph
data:
config: |
[global]
bluestore_min_alloc_size = 16384

2、 I installed rook. 
kubectl create -f common.yaml
kubectl create -f operator.yaml

then  kubectl create -f ceph.yaml  . after i apply the ceph.yaml, i created the cluster.
 kubectl create -f cluster.yaml

3、after the cluster created, i uploaded 4-10KB files into bucket. and i found that it didn't work well.it still wasted much more space

**Comment by travisn** [comment]
@ijackcy Can you check as described in [this comment](https://github.com/rook/rook/issues/6140#issuecomment-679392189) that the expected setting is actually applied to the OSD? Sometimes there can be a formatting issue in the configmap that prevents it from being picked up.

**Comment by ijackcy** [comment]
@travisn   i checked the osd config in the toolbox. 

[root@rook-ceph-tools-75f575489-4q995 /]# ceph config show osd.1   
NAME                               VALUE                                                                                                                               SOURCE   OVERRIDES   IGNORES 
bluestore_min_alloc_size           16384                                                                                                                               file                         
bluestore_min_alloc_size_hdd       16384                                                                                                                                file                         
crush_location                     root=default host=slaver001                                                                                                         cmdline                      
daemonize                          false                                                                                                                               override                     
keyring                            $osd_data/keyring                                                                                                                   default                      
leveldb_log                                                                                                                                                            default         



and i enter the osd.0 pod , i want to check the bluestore_stored and bluestore_allocated, but nothing.

[root@rook-ceph-osd-0-5bb95674f8-j459r /]# ceph --admin-daemon /var/run/ceph/ceph-osd.0.asok perf dump
{}
[root@rook-ceph-osd-0-5bb95674f8-j459r /]# ceph --admin-daemon /var/run/ceph/ceph-osd.0.asok perf dump
{}

**Comment by ijackcy** [comment]
@travisn thank you!

I refer to your method and try it. and modify bluestore_min_alloc_size 8K , and i upload files(The size of a single file is about 3-5kB) into the bucket.

[root@rook-ceph-osd-1-6c66676575-btcl6 /]# ceph daemon osd.1 config show  | grep bluestore_min
    "bluestore_min_alloc_size": "8192",
    "bluestore_min_alloc_size_hdd": "8192",
    "bluestore_min_alloc_size_ssd": "4096",

[root@rook-ceph-osd-1-6c66676575-btcl6 /]# ceph daemon osd.1 perf dump | grep bluestore_      
        "bluestore_allocated": 66519040,
        "bluestore_stored": 3847435,

and  i found that "bluestore_allocated"   "bluestore_stored" was not right. In theory, the two values should be close.

**Comment by TonyBogdanov** [comment]
I'm experiencing the same issue. `bluestore_min_alloc_size` is set to `1024` in the config-map, yet files smaller than 64KB use exactly 64KB of space, whereas I would expect them to use 1KB (or more if the file is larger). Am I missing some other setting?

**Comment by ijackcy** [comment]
I tried it many times,   i specify bluestore_ min_ alloc_ size = 8192 in the ceph.conf overrides, and i created the rook-config-override ConfigMap before the cluster is started.

then i created the cluster. i came into the osd pod, checked the bluestore_min_alloc_size_hdd = 8192

ceph daemon osd.1 config show | grep bluestore_min

But when I wrote some small files(1-3KB), the storage space consumed is much larger than the actual size of the files,and many space was wasted.

**Comment by TonyBogdanov** [comment]
> I tried it many times, i specify bluestore_ min_ alloc_ size = 8192 in the ceph.conf overrides, and i created the rook-config-override ConfigMap before the cluster is started.
> 
> then i created the cluster. i came into the osd pod, checked the bluestore_min_alloc_size_hdd = 8192
> 
> ceph daemon osd.1 config show | grep bluestore_min
> 
> But when I wrote some small files(1-3KB), the storage space consumed is much larger than the actual size of the files,and many space was wasted.

I believe it is useless to try this and just accept it is what it is. I wasted a week now trying to get it working and couldn't, then last night I accidentally found a hardcoded constant in the Linux kernel header files setting 64K as the smallest possible value for this setting. I could be wrong, of course, but it is consistent with my and other people's experiences.

What I ended up doing in my case is use erasure coding with 3 OSDs instead of replica 3. It uses 33% less space on the expense of some extra CPU & RAM.

**Comment by satoru-takeuchi** [comment]
> last night I accidentally found a hardcoded constant in the Linux kernel header files setting 64K as the smallest possible value for this setting

@TonyBogdanov Could you tell me where it is?

**Comment by TonyBogdanov** [comment]
@satoru-takeuchi I don't know if it's anything official, but I found it via a Google search and thought it was related: https://elixir.bootlin.com/linux/latest/source/include/linux/ceph/ceph_fs.h#L76 (it is also available on various other sites & kernel versions).

**Comment by satoru-takeuchi** [comment]
@TonyBogdanov Thanks, I'll take a look at it.

I wonder why everything go well in @ijackcy 's ceph-deploy environment and it's not in Rook, even though the configuration seems to be reflected properly.

**Comment by TonyBogdanov** [comment]
@satoru-takeuchi He did say he was using Ceph v14, whereas currently Rook uses Ceph v15.2.4. I have a feeling this is indeed a Ceph issue, not a Rook one. In my case the settings from the config map *are* properly reflected in Ceph, they just don't work.

Do you know which most recent release of Rook uses Ceph v14 so I can test it?

**Comment by ijackcy** [comment]
I installed Ceph v14 through CEPH-deploy and modified / etc / CEPH/ ceph.conf.bluestore_min_alloc_size_hdd = 4096
and I wrote some small files(1-3KB), The storage space consumed is close to the actual size of the file.
https://docs.ceph.com/docs/hammer/start/quick-start-preflight/#ceph-deploy-setup

I installed Ceph v15 through cephadm and modified / etc / CEPH/ ceph.conf.bluestore_min_alloc_size_hdd = 4096
and I wrote some small files(1-3KB), Also the storage space consumed is close to the actual size of the file.
cephadm used docker to create ceph.
https://docs.ceph.com/docs/master/cephadm/install/
 i came into the OSD POD, saved the config of osd.1
ceph daemon osd.1 config show > cephadm-config.txt

also in rook 1.4.0, i came into the OSD POD, saved the config of osd.1
ceph daemon osd.1 config show > rook-config.txt

Then, i compared the two configuration files( rook-config.txt and cephadm-config.txt )and found that all other parameters are the same, only OSD_MEMORY_TARGET this parameter is different.

For all the above tests, I used the same node servers and master server.

I'm very confused.

**Comment by satoru-takeuchi** [comment]
> Then, i compared the two configuration files( rook-config.txt and cephadm-config.txt )and found that all other parameters
> are the same, only OSD_MEMORY_TARGET this parameter is different.

@ijackcy  , it's interesting. This parameter is controlled under Rook (see #6140 ). I'll try to investigate this problem by considering this parameter.

> Do you know which most recent release of Rook uses Ceph v14 so I can test it?
> 
> EDIT Just tested it on a brand new cluster and the issue also exists on Rook release-1.3 with Ceph 14.2.10. Don't have
> time right now to test it without Rook though.

@TonyBogdanov , thank you for your effort!

**Comment by ijackcy** [comment]
My environment is : 1 master, 3 slave nodes, each node has 2 raw disks for OSD.

Rook and cephadm are installed in the test environment at the same time, and the version of hook is 1.4.0. Since CEPH official website does not recommend ceph-deploy mode to deploy CEPH, I choose cephadm mode to deploy, and cephadm uses docker to install CEPH. Rook uses 3 raw disks for OSDS, cephadm uses 3 raw disks for osds.

In rook, 
kubectl create -f common.yaml

kubectl create -f operator.yaml

kubectl create -f ceph-config.yaml

 cat ceph-config.yaml

apiVersion: v1

kind: ConfigMap

metadata:

  name: rook-config-override

  namespace: rook-ceph

data:

  config: |

    [global]

    osd pool default size = 2

    [osd]

    bluestore min alloc size = 4096

    bluestore min alloc size hdd = 4096 


then i created the rook cluster.

I set up object storage in Rook and cephadm, upload the same folder, containing 10000 small files. The actual size of the file is about 2-16kB, and the total size of the file is about 114.8MB. Then I use the command rados df to observe the actual occupied space, and find that Rook occupies 1.9 GB, 3 copies; but cephadm occupies 403 MB, 3 copies.

1、cephadm rados df

root@slaver001:/# rados df

POOL_NAME USED  OBJECTS  CLONES  COPIES  MISSING_ON_PRIMARY  UNFOUND  DEGRADED  RD_OPS       RD  WR_OPS       WR  USED COMPR  UNDER COMPR

.rgw.root 144 KiB       13       0      39                   0        0         0      75   75 KiB      21   16 KiB         0 B          0 B

device_health_metrics  0 B        0       0       0                   0        0         0       0      0 B       0      0 B         0 B          0 B

my-store.rgw.buckets.data 403 MiB    10299       0   30897                   0        0         0       0      0 B  123534  120 MiB         0 B          0 B



cephadm version:

root@slaver001:/# ceph -v

ceph version 15.2.4 (7447c15c6ff58d7fce91843b705a268a1917325c) octopus (stable)

2、cephadm osd config bluestore_min_alloc_size_hdd = 4096

root@slaver001:/# ceph daemon osd.0 config show  | grep bluestore_min

    "bluestore_min_alloc_size": "4096",

    "bluestore_min_alloc_size_hdd": "4096",

    "bluestore_min_alloc_size_ssd": "4096",

root@slaver001:/# 

root@slaver001:/# ceph daemon osd.0 perf dump | grep bluestore_      

        "bluestore_allocated": 144240640,

        "bluestore_stored": 121886986,


and bluestore_allocated is close to bluestore_stored.



3.rook rados df

[root@rook-ceph-tools-75b55dbcb9-42f75 /]# rados df

POOL_NAME  USED  OBJECTS  CLONES  COPIES  MISSING_ON_PRIMARY  UNFOUND  DEGRADED  RD_OPS       RD  WR_OPS       WR  USED COMPR  UNDER COMPR

.rgw.root  2.8 MiB       16       0      48                   0        0         0   55202   54 MiB      28   21 KiB         0 B          0 B

device_health_metrics            0 B        0       0       0                   0        0         0       0      0 B       0      0 B         0 B          0 B

my-store.rgw.buckets.data    1.9 GiB    10299       0   30897                   0        0         0   16072   18 MiB  151086  122 MiB         0 B          0 B



4. rook ceph version

[root@rook-ceph-osd-0-86c9fdbf5d-rkn95 /]# ceph -v

ceph version 15.2.4 (7447c15c6ff58d7fce91843b705a268a1917325c) octopus (stable)



5.rook osd config 

[root@rook-ceph-osd-0-86c9fdbf5d-rkn95 /]# ceph daemon osd.0 config show  | grep bluestore_min

    "bluestore_min_alloc_size": "4096",

    "bluestore_min_alloc_size_hdd": "4096",

    "bluestore_min_alloc_size_ssd": "4096",



[root@rook-ceph-osd-0-86c9fdbf5d-rkn95 /]# 

[root@rook-ceph-osd-0-86c9fdbf5d-rkn95 /]# 

[root@rook-ceph-osd-0-86c9fdbf5d-rkn95 /]# ceph daemon osd.0 perf dump | grep bluestore_      

        "bluestore_allocated": 750649344,

        "bluestore_stored": 128021128,

but bluestore_allocated is much bigger than bluestore_stored.


now i am confused, If you need any test, please feel free to contact me.

**Comment by satoru-takeuchi** [comment]
@TonyBogdanov @travisn 

I discussed this issue with @ijackcy in email yesterday. Here is the result.

I suspect the root cause is in Ceph and the trigger is changing osd_memory_target from the default value (4294967296). I asked him to confirm whether cephadm cluster works fine with changing this parameter. If it doesn't, the root cause is apparently not in Rook, but in Ceph.

In any case, I'll submit an Ceph issue after getting the test result.

**Comment by satoru-takeuchi** [comment]
> I suspect the root cause is in Ceph and the trigger is changing osd_memory_target from the default value (4294967296).
> I asked him to confirm whether cephadm cluster works fine with changing this parameter.

A cephadm cluster with changing osd_memory_target parameter worked fine. Here is the result (the further discussion will be not in email, but in this issue.)

===
I redeployed Ceph using Cephadm and specified the ceph. conf file.
cephadm bootstrap --mon-ip 192.168.100.21 --config /etc/ceph.conf 
root@slaver001:/etc/ceph# cat ../ceph.conf 
[osd]
        bluestore_min_alloc_size = 4096
        bluestore_min_alloc_size_hdd = 4096
        bluestore_min_alloc_size_ssd = 4096
        osd_memory_target = 6605268582

then i came into the osd pod,check the osd_memory_target =6605268582, 

[root@slaver002 /]# ceph daemon osd.1 config show  | grep osd_memory_target
    "osd_memory_target": "6605268582",
    "osd_memory_target_cgroup_limit_ratio": "0.800000",


[root@slaver002 /]# ceph daemon osd.1 config show  | grep bluestore_min       

    "bluestore_min_alloc_size": "4096",

    "bluestore_min_alloc_size_hdd": "4096",

    "bluestore_min_alloc_size_ssd": "4096",



and i  upload the same folder, containing 10000 small files. The actual size of the file is about 2-16kB, and the total size of the file is about 114.8MB. Then I use the command rados df to observe the actual occupied space, cephadm still occupies 403 MB, 3 copies.

root@slaver001:/# rados df
POOL_NAME                        USED  OBJECTS  CLONES  COPIES  MISSING_ON_PRIMARY  UNFOUND  DEGRADED  RD_OPS       RD  WR_OPS       WR  USED COMPR  UNDER COMPR
.rgw.root                     144 KiB       13       0      39                   0        0         0      76   76 KiB      21   16 KiB         0 B          0 B
cn-hangzho.rgw.buckets.data   403 MiB    10299       0   30897                   0        0         0       0      0 B  123534  120 MiB         0 B          0 B


[root@slaver002 /]# ceph daemon osd.1 perf dump | grep bluestore_   
        "bluestore_allocated": 143962112,
        "bluestore_stored": 121799448,
also bluestore_allocated is close to bluestore_stored.

**Comment by ashangit** [comment]
Hi,

The issue seems to be on the rook-ceph-osd-prepare job which doesn't take in account config from the rook-config-override ConfigMap. This leads to the ceph-osd -mkfs command to be run without the appropriate bluestore_min_alloc_size/bluestore_min_alloc_size_hdd, so initializing the bluestore with default config of 64Kb for hdd.

Here is a test I've done to confirm that creating 100k objects with 1bytes of data each result to:
- on osd being created by rook 1.3.9 with bluestore_min_alloc_size_hdd = 4096:
```
[vagrant@localhost ~]$ kubectl  exec -ti rook-ceph-tools-68fdbcb766-8bnj9 -- ceph osd df
ID CLASS WEIGHT  REWEIGHT SIZE   RAW USE DATA    OMAP   META     AVAIL  %USE  VAR  PGS STATUS 
 0   hdd 0.02930  1.00000 30 GiB 3.1 GiB 2.1 GiB    0 B    1 GiB 27 GiB 10.43 1.03  61     up 
 1   hdd 0.02930  1.00000 30 GiB 3.1 GiB 2.1 GiB 12 MiB 1012 MiB 27 GiB 10.35 1.02  68     up 
 2   hdd 0.02930  1.00000 30 GiB 2.9 GiB 1.9 GiB    0 B    1 GiB 27 GiB  9.74 0.96  47     up 
                    TOTAL 90 GiB 9.2 GiB 6.2 GiB 12 MiB  3.0 GiB 81 GiB 10.17  
```

- on osd being created by an ugly patched release of rook 1.3.9 (https://github.com/ashangit/rook/commit/e3a76020276908c3e9ede9b882e2d95fe4ff3aad) with bluestore_min_alloc_size_hdd = 4096:
```
[vagrant@localhost data]$ kubectl exec -ti rook-ceph-tools-68fdbcb766-jrzzp  -- ceph osd df
ID CLASS WEIGHT  REWEIGHT SIZE   RAW USE DATA    OMAP   META     AVAIL  %USE VAR  PGS STATUS 
 0   hdd 0.02930  1.00000 30 GiB 1.1 GiB 111 MiB    0 B    1 GiB 29 GiB 3.69 0.98  26     up 
 1   hdd 0.02930  1.00000 30 GiB 1.1 GiB 147 MiB 11 MiB 1013 MiB 29 GiB 3.81 1.01  34     up 
 2   hdd 0.02930  1.00000 30 GiB 1.1 GiB 134 MiB    0 B    1 GiB 29 GiB 3.77 1.00  20     up 
                    TOTAL 90 GiB 3.4 GiB 392 MiB 11 MiB  3.0 GiB 87 GiB 3.76  
```

**Comment by TonyBogdanov** [comment]
@ashangit Excellent find! It makes complete sense to me. I'm up to helping with fixing this if someone can point me in the right direction :)

**Comment by satoru-takeuchi** [comment]
@ashangit Great work!

Do you plan to create a PR? If not, I'll do it.

I guess the PR will inject `rook-config-override` to `rook-ceph-osd` job.

https://github.com/rook/rook/blob/master/pkg/operator/ceph/cluster/osd/provision_spec.go#L225

**Comment by ashangit** [comment]
@satoru-takeuchi I will check how I can merge both config as the default one as some important config in it.
probably reading the `rook-config-override` configmap and use iniFile API to merge both of them when generating the job conf

**Comment by cehoffman** [comment]
This problem appears to still exist in Rook v1.11.9 with Ceph 17.2.6. I have tried many variations of setting `bluestore_min_alloc_size` where I was preferring to have this only apply to our HDD OSDs because they are used very large files only and we have NVMe for the many small file usecase. It isn't clear to me where this setting lives in INI as i've seen it in both `global` and `osd` (base on https://github.com/rook/rook/issues/6032#issuecomment-690006664) sections and [official documentation](https://docs.ceph.com/en/quincy/rados/configuration/bluestore-config-ref/#confval-bluestore_min_alloc_size) doesn't indicate. After cycling an OSD out and back into the cluster, configuration still shows the default `0` for  `bluestore_min_alloc_size` and `4096` for `bluestore_min_alloc_size_hdd` on the new OSD.

```ini
[global]
public network = fd02::/64
cluster network = fd01::/64
public addr = ""
cluster addr = ""

bluestore_min_alloc_size = 65536
bluestore_min_alloc_size_hdd = 65536

[osd]
bluestore_min_alloc_size = 65536
bluestore_min_alloc_size_hdd = 65536
```

Configuration confirmed for the OSD with

```sh
$ ceph config get osd.5 bluestore_min_alloc_size
0
$ ceph config get osd.5 bluestore_min_alloc_size_hdd
4096
```

**Comment by cehoffman** [comment]
Yes, this is using the ceph.conf override with the configmap. The network settings for front and back addr do come through as I included in the sample. The bluestore settings are not reflected in the final OSD.

I will need to wait a bit still before I can try again. The tip on `unset CEPH_ARGS` is very helpful as I wasn't able to get the osd pod `ceph daemon` command to work previously and didn't know what to do to fix it.

I learned a bit more about the ceph.conf organization after doing a bunch of tuning on recovery and backfill settings, so I see now that for what I was trying it didn't matter which section as either would have worked. Thanks for the tip.
